File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // Build adjacency list
        vector<vector<int>> adj(numCourses);
        for (auto& pre : prerequisites) {
            adj[pre[1]].push_back(pre[0]);
        }

        // 0=unvisited, 1=visiting, 2=visited
        vector<int> visited(numCourses, 0);

        // DFS function
        function<bool(int)> dfs = [&](int node) {
            if (visited[node] == 1) return false; // cycle
            if (visited[node] == 2) return true;  // already processed

            visited[node] = 1; // mark as visiting
            for (int nei : adj[node]) {
                if (!dfs(nei)) return false;
            }
            visited[node] = 2; // mark as processed
            return true;
        };

        // Check all courses
        for (int i = 0; i < numCourses; ++i) {
            if (!dfs(i)) return false;
        }
        return true;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
from typing import List
from collections import deque

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # Build adjacency list and in-degree count
        graph = [[] for _ in range(numCourses)]
        in_degree = [0] * numCourses

        for course, prereq in prerequisites:
            graph[prereq].append(course)
            in_degree[course] += 1

        # Start with all courses that have no prerequisites
        queue = deque([course for course in range(numCourses) if in_degree[course] == 0])
        completed = 0

        while queue:
            current = queue.popleft()
            completed += 1

            for next_course in graph[current]:
                in_degree[next_course] -= 1
                if in_degree[next_course] == 0:
                    queue.append(next_course)

        # If we completed all courses, there's no cycle
        return completed == numCourses  